File: src/flow.cpp
// FLOW
// Attila Bagyoni, 2018
// This software is public domain. Do whatever you want with it.

#include <cstdlib>
#include <cmath>

#include "lib/main.h"
#include "lib/perlin.h"

uint16_t const NUM_PARTICLES = 5000;
uint16_t const PATH_LEN = 300;

double flow[WID * HEI]; // TODO refactor!!!
double curvature;
double frequency;
int8_t signum;

void make_single_path() {
    double x = rand() % bytes->width;
    double y = rand() % bytes->height;
    int i = 0;
    while (i < PATH_LEN && x > 0 && x < bytes->width && y > 0 && y < HEI) {
        int xi = x;
        int yi = y;
        flow[(bytes->width * yi + xi)] += ((double) (PATH_LEN - i)) / PATH_LEN;
        double angle = 2 * M_PI * (fractal(x, y, frequency, 6) - 0.5) * curvature;
        x += cos(angle);
        y += sin(angle);
        i++;
    }
}

void draw() {
    generate_noise();
    signum = (rand() & 0x1) ? 1 : -1;
    curvature = 0.3 + 0.1 * ((float) (rand() & 0xFF)) / 0xFF;
    frequency = 0.003 + 0.003 * ((float) (rand() & 0xFF)) / 0xFF;
    int i;
    for (i = 0; i < NUM_PARTICLES; i++) {
        make_single_path();
    }
    double max;
    for (i = 0; i < bytes->width * bytes->height; i++) {
        max = flow[i] > max ? flow[i] : max;
    }
    for (int x = 0; x < bytes->width; x++) {
        for (int y = 0; y < bytes->height; y++) {
            auto val = (uint8_t) (256.0 + (double) signum * (55.0 + 200.0 * flow[y * bytes->width + x] / (double) max));
            bytes->setR(x, y, val);
            bytes->setG(x, y, val);
            bytes->setB(x, y, val);
        }
    }
}

File: wallpapers/src/algorithms/flow.rs
use crate::utils::perlin::Perlin;
use image::{Rgb, RgbImage};
use rand::Rng;
use std::f64::consts::PI;
use crate::algorithms::Algorithm;

/// Flow
/// Based on the FLOW algorithm implementation by Attila Bagyoni, 2018
/// https://github.com/bagyoni/procedural-wallpapers

pub struct Flow {
    num_particles: u32,
    path_len: u32,
    curvature: f64,
    frequency: f64,
    signum: bool,
}

impl Default for Flow {
    fn default() -> Self {
        Flow {
            num_particles: 5000,
            path_len: 300,
            curvature: 0.0,
            frequency: 0.0,
            signum: true,
        }
    }
}

impl Flow {
    /// Get the signum as integer
    fn signum(&self) -> i8 {
        match self.signum {
            true => 1,
            false => -1,
        }
    }
    fn make_single_path(
        &self,
        rng: &mut impl Rng,
        perlin: &Perlin,
        flow: &mut [Vec<f64>],
        img: &RgbImage,
    ) -> Result<(), String> {
        let mut x: f64 = rng.gen_range(0..img.width()) as f64;
        let mut y: f64 = rng.gen_range(0..img.height()) as f64;
        let mut i: u32 = 0;
        while i < self.path_len
            && x > 0.0
            && x < img.width() as f64
            && y > 0.0
            && y < img.height() as f64
        {
            flow[x as usize][y as usize] += ((self.path_len - i) as f64) / self.path_len as f64;
            let angle: f64 =
                2.0 * PI * (perlin.fractal(x, y, self.frequency, 6)? - 0.5) * self.curvature;
            x += angle.cos();
            y += angle.sin();
            i += 1;
        }
        Ok(())
    }
}

impl<R: Rng> Algorithm<R> for Flow {
    fn build(&mut self, rng: &mut R, img: &mut RgbImage) -> Result<(), String> {
        let mut flow = vec![vec![0.0; img.height() as usize]; img.width() as usize];
        let mut perlin: Perlin = Perlin::new(img.width() as usize, img.height() as usize);
        perlin.regenerate_noise(rng);
        self.signum = rng.gen_bool(0.5);
        self.curvature = 0.3 + 0.1 * ((rng.gen::<u8>() as f64) / 0xff as f64);
        self.frequency = 0.003 + 0.003 * ((rng.gen::<u8>() as f64) / 0xff as f64);
        for _ in 0..self.num_particles {
            self.make_single_path(rng, &perlin, &mut flow, img)?;
        }
        let max: f64 = flow
            .iter()
            .flatten()
            .fold(f64::NEG_INFINITY, |prev, curr| prev.max(*curr));
        for (x, y, pixel) in img.enumerate_pixels_mut() {
            let val: u8 = (256.0
                + (self.signum() as f64) * (55.0 + 200.0 * flow[x as usize][y as usize] / max))
                as u8;
            *pixel = Rgb([val, val, val]);
        }
        Ok(())
    }
}
